File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Go/17-分账/DeleteReceiver.go
package main

import (
	"bytes"
	"demo/wxpay_utility" // 引用微信支付工具库，参考 https://pay.weixin.qq.com/doc/v3/partner/4015119446
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
)

func main() {
	config, err := wxpay_utility.CreateMchConfig(
		"YOUR_MCHID",
		"YOUR_CERT_SERIAL_NO",
		"/path/to/apiclient_key.pem",
		"YOUR_PUB_KEY_ID",
		"/path/to/wxp_pub.pem",
	)
	if err != nil {
		fmt.Println(err)
		return
	}

	request := &DeleteReceiverRequest{
		SubMchid: wxpay_utility.String("1900000109"),
		Appid:    wxpay_utility.String("wxYOUR_APPID"),
		SubAppid: wxpay_utility.String("wx8888888888888889"),
		Type:     RECEIVERTYPE_MERCHANT_ID.Ptr(),
		Account:  wxpay_utility.String("1900000109"),
	}

	response, err := DeleteReceiver(config, request)
	if err != nil {
		fmt.Printf("请求失败: %+v\n", err)
		return
	}

	fmt.Printf("请求成功: %+v\n", response)
}

func DeleteReceiver(config *wxpay_utility.MchConfig, request *DeleteReceiverRequest) (response *DeleteReceiverResponse, err error) {
	const (
		host   = "https://api.mch.weixin.qq.com"
		method = "POST"
		path   = "/v3/profitsharing/receivers/delete"
	)

	reqUrl, err := url.Parse(fmt.Sprintf("%s%s", host, path))
	if err != nil {
		return nil, err
	}
	reqBody, err := json.Marshal(request)
	if err != nil {
		return nil, err
	}
	httpRequest, err := http.NewRequest(method, reqUrl.String(), bytes.NewReader(reqBody))
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Accept", "application/json")
	httpRequest.Header.Set("Wechatpay-Serial", config.WechatPayPublicKeyId())
	httpRequest.Header.Set("Content-Type", "application/json")
	authorization, err := wxpay_utility.BuildAuthorization(config.MchId(), config.CertificateSerialNo(), config.PrivateKey(), method, reqUrl.RequestURI(), reqBody)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Authorization", authorization)

	client := &http.Client{}
	httpResponse, err := client.Do(httpRequest)
	if err != nil {
		return nil, err
	}
	respBody, err := wxpay_utility.ExtractResponseBody(httpResponse)
	if err != nil {
		return nil, err
	}
	if httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300 {
		err = wxpay_utility.ValidateResponse(config.WechatPayPublicKeyId(), config.WechatPayPublicKey(), &httpResponse.Header, respBody)
		if err != nil {
			return nil, err
		}
		response := &DeleteReceiverResponse{}
		if err := json.Unmarshal(respBody, response); err != nil {
			return nil, err
		}
		return response, nil
	} else {
		return nil, wxpay_utility.NewApiException(httpResponse.StatusCode, httpResponse.Header, respBody)
	}
}

type DeleteReceiverRequest struct {
	SubMchid *string       `json:"sub_mchid,omitempty"`
	Appid    *string       `json:"appid,omitempty"`
	SubAppid *string       `json:"sub_appid,omitempty"`
	Type     *ReceiverType `json:"type,omitempty"`
	Account  *string       `json:"account,omitempty"`
}

type DeleteReceiverResponse struct {
	Type    *ReceiverType `json:"type,omitempty"`
	Account *string       `json:"account,omitempty"`
}

type ReceiverType string

func (e ReceiverType) Ptr() *ReceiverType { return &e }

const (
	RECEIVERTYPE_MERCHANT_ID         ReceiverType = "MERCHANT_ID"
	RECEIVERTYPE_PERSONAL_OPENID     ReceiverType = "PERSONAL_OPENID"
	RECEIVERTYPE_PERSONAL_SUB_OPENID ReceiverType = "PERSONAL_SUB_OPENID"
)

File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Java/17-分账/DeleteReceiver.java
package com.java.demo;

import com.java.utils.WXPayUtility; // 引用微信支付工具库，参考：https://pay.weixin.qq.com/doc/v3/partner/4014985777

import com.google.gson.annotations.SerializedName;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.PrivateKey;
import java.security.PublicKey;

/**
 * 删除分账接收方（服务商模式）
 */
public class DeleteReceiver {
  private static String HOST = "https://api.mch.weixin.qq.com";
  private static String METHOD = "POST";
  private static String PATH = "/v3/profitsharing/receivers/delete";

  public static void main(String[] args) {
    DeleteReceiver client = new DeleteReceiver(
      "YOUR_MCHID",
      "YOUR_CERT_SERIAL_NO",
      "/path/to/apiclient_key.pem",
      "YOUR_PUB_KEY_ID",
      "/path/to/wxp_pub.pem"
    );

    DeleteReceiverRequest request = new DeleteReceiverRequest();
    request.subMchid = "1900000109";
    request.appid = "wxYOUR_APPID";
    request.subAppid = "wx8888888888888889";
    request.type = ReceiverType.MERCHANT_ID;
    request.account = "1900000109";
    try {
      DeleteReceiverResponse response = client.run(request);
      System.out.println(response);
    } catch (WXPayUtility.ApiException e) {
      e.printStackTrace();
    }
  }

  public DeleteReceiverResponse run(DeleteReceiverRequest request) {
    String uri = PATH;
    String reqBody = WXPayUtility.toJson(request);

    Request.Builder reqBuilder = new Request.Builder().url(HOST + uri);
    reqBuilder.addHeader("Accept", "application/json");
    reqBuilder.addHeader("Wechatpay-Serial", wechatPayPublicKeyId);
    reqBuilder.addHeader("Authorization", WXPayUtility.buildAuthorization(mchid, certificateSerialNo, privateKey, METHOD, uri, reqBody));
    reqBuilder.addHeader("Content-Type", "application/json");
    RequestBody requestBody = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), reqBody);
    reqBuilder.method(METHOD, requestBody);
    Request httpRequest = reqBuilder.build();

    OkHttpClient client = new OkHttpClient.Builder().build();
    try (Response httpResponse = client.newCall(httpRequest).execute()) {
      String respBody = WXPayUtility.extractBody(httpResponse);
      if (httpResponse.code() >= 200 && httpResponse.code() < 300) {
        WXPayUtility.validateResponse(this.wechatPayPublicKeyId, this.wechatPayPublicKey,
            httpResponse.headers(), respBody);
        return WXPayUtility.fromJson(respBody, DeleteReceiverResponse.class);
      } else {
        throw new WXPayUtility.ApiException(httpResponse.code(), respBody, httpResponse.headers());
      }
    } catch (IOException e) {
      throw new UncheckedIOException("Sending request to " + uri + " failed.", e);
    }
  }

  private final String mchid;
  private final String certificateSerialNo;
  private final PrivateKey privateKey;
  private final String wechatPayPublicKeyId;
  private final PublicKey wechatPayPublicKey;

  public DeleteReceiver(String mchid, String certificateSerialNo, String privateKeyFilePath, String wechatPayPublicKeyId, String wechatPayPublicKeyFilePath) {
    this.mchid = mchid;
    this.certificateSerialNo = certificateSerialNo;
    this.privateKey = WXPayUtility.loadPrivateKeyFromPath(privateKeyFilePath);
    this.wechatPayPublicKeyId = wechatPayPublicKeyId;
    this.wechatPayPublicKey = WXPayUtility.loadPublicKeyFromPath(wechatPayPublicKeyFilePath);
  }

  public static class DeleteReceiverRequest {
    @SerializedName("sub_mchid")
    public String subMchid;

    @SerializedName("appid")
    public String appid;

    @SerializedName("sub_appid")
    public String subAppid;

    @SerializedName("type")
    public ReceiverType type;

    @SerializedName("account")
    public String account;
  }

  public static class DeleteReceiverResponse {
    @SerializedName("type")
    public ReceiverType type;

    @SerializedName("account")
    public String account;
  }

  public enum ReceiverType {
    @SerializedName("MERCHANT_ID")
    MERCHANT_ID,
    @SerializedName("PERSONAL_OPENID")
    PERSONAL_OPENID,
    @SerializedName("PERSONAL_SUB_OPENID")
    PERSONAL_SUB_OPENID
  }

}
